File: Clocktower.Server/Data/GamePerspective.cs
namespace Clocktower.Server.Data;

public record GamePerspective(string Id, string UserId, string GuildId, GameUser CreatedBy, DateTime CreatedDate) : IIdentifiable
{
    public IReadOnlyList<GameUser> Users { get; init; } = [];
    public IEnumerable<GameUser> Players => Users.Where(o => o.UserType == UserType.Player);
    public IEnumerable<GameUser> StoryTellers => Users.Where(o => o.UserType == UserType.StoryTeller);
    public IEnumerable<GameUser> Spectators => Users.Where(o => o.UserType == UserType.Spectator);
    public GameTime GameTime { get; init; }
    public Script? Script { get; init; }

    public RoleDistribution? DefaultRoleDistributions
    {
        get
        {
            int playerCount = Players.Count();
            if (playerCount < 5) return null;

            const int demons = 1;
            int minions = (playerCount - 7) / 3 + 1;
            int outsiders = playerCount switch
            {
                5 => 0,
                6 => 1,
                _ => (playerCount - 7) % 3
            };
            int townsfolk = playerCount - outsiders - minions - demons;

            return new RoleDistribution(townsfolk, outsiders, minions, demons);
        }
    }
}

public record RoleDistribution(int Townsfolk, int Outsiders, int Minions, int Demons);

File: Clocktower-UI/src/types/gamePerspective.ts
import type {ClocktowerServerDataGamePerspective} from "@/api";
import {type GameUser, mapToGameUser} from "@/types/gameUser.ts";
import {UserType} from "@/types/userType.ts";

export type GamePerspective = {
    id: string;
    userId: string;
    guildId: string,
    users: GameUser[];
    players: GameUser[];
    spectators: GameUser[];
    storyTellers: GameUser[];
    createdDate: Date,
    createdBy: GameUser
}

export function mapToGamePerspective(apiPerspective: ClocktowerServerDataGamePerspective): GamePerspective {
    if (!apiPerspective.id || !apiPerspective.userId || !apiPerspective.guildId) {
        throw new Error('Invalid API response: missing required fields');
    }
    return {
        id: apiPerspective.id,
        userId: apiPerspective.userId,
        guildId: apiPerspective.guildId,
        users: (apiPerspective.users ?? []).map(mapToGameUser),
        players: (apiPerspective.players ?? []).map(mapToGameUser),
        spectators: (apiPerspective.spectators ?? []).map(mapToGameUser),
        storyTellers: (apiPerspective.storyTellers ?? []).map(mapToGameUser),
        createdDate: apiPerspective.createdDate ?? new Date(),
        createdBy: apiPerspective.createdBy ? mapToGameUser(apiPerspective.createdBy) : {id: '', isPlaying: false, userType: UserType.Unknown}
    };
}